pkg/util/docker/dockerfile/dockerfile.go

```diff
@@ -0,0 +1,58 @@
+package dockerfile
+
+import (
+	"fmt"
+	"strings"
+
+	"github.com/docker/docker/builder/parser"
+)
+
+// ParseTreeToDockerfile takes a Dockerfile AST node, as generated by
+// parser.Parse, and returns a new, equivalent, Dockerfile.
+func ParseTreeToDockerfile(node *parser.Node) []byte {
+	buf := []byte(node.Original)
+	for _, child := range node.Children {
+		buf = append(buf, ParseTreeToDockerfile(child)...)
+	}
+	// Append a line break when needed.
+	if len(buf) > 0 && buf[len(buf)-1] != '\n' {
+		buf = append(buf, '\n')
+	}
+	return buf
+}
+
+// FindAll returns the indices of all children of node such that
+// node.Children[i].Value == command.
+func FindAll(node *parser.Node, command string) []int {
+	var indices []int
+	for i, child := range node.Children {
+		if child != nil && child.Value == command {
+			indices = append(indices, i)
+		}
+	}
+	return indices
+}
+
+// InsertInstructions inserts instructions as the pos-th child of node, moving
+// other children as necessary. The instructions should be valid Dockerfile
+// instructions.
+func InsertInstructions(node *parser.Node, pos int, instructions string) error {
+	if node == nil {
+		return fmt.Errorf("cannot insert instructions in a nil node")
+	}
+	if pos < 0 || pos > len(node.Children) {
+		return fmt.Errorf("invalid pos: %d", pos)
```

Comment:
`pos` refers to the argument name. But the message will be better in my next push.... :-)